PIVX-Project/PIVX
# Sapling transaction primitive data.

Work coming from #1798. Decoupled commits related to the Sapling transaction primitive data and the new signature hash to make the review a little bit easier there. Plus, squashed some of them.

List of commits taken from #1798:

Primitive data:

* Add serialization for primitive boost::optional<T>. —> 659cc558
* Transaction: Implement BaseOutPoint class.  —> f17b23c7
* Backport zkSNARK compression. —> 5dbc3c9
* Basic Sapling transaction primitive data --> c49594f
* Add serialization for std:list objects. —> 7cc57f7
* sapling_transaction.h migrated to the new serialization —> dbb618f6
* Fix sighash_tests —> e1e33163 
* Transaction GetShieldedValueIn & GetValueOut back port —> 0fbe0236
* BugFix in GetShieldedValueIn —> 57ee1bc6
* Missing SaplingOutPoint::ToString() method implemented —> e1fe5bf7
* Transaction: add sapling data to toString —> 26ce38b5
* make SaplingTxData Optional in transaction primitive  — PARTIALLY  —> c3b1aedc
* Cleaning compiler warnings — PARTIALLY —> 644c381f
* transaction: fix sapling version to 2. —> 8fc595f5
* Transaction: GetDustThreshold method implemented. —> dadf199c.
* Transaction:hasSaplingData check is enough checking for vShieldedOutput and vShieldedSpend emptiness. —> 57d3236f

Signature hash:

* Introduce SignatureHash personalization updates —> 550217f7
* Enable sigversion for Sapling transactions. —> c04abe1d 
* Sapling signature hash (Custom version of ZIP 243) —> eac206e9
* Not try to serialize empty fields in Sapling signature hash  --->  ce5a467
